File: packages/verify/src/modules/challenge/challenge.service.ts
import blsInit from 'bls-signatures'

type BlsModule = Awaited<ReturnType<typeof blsInit>>

let blsReady: Promise<BlsModule> | undefined

async function ensureBls(): Promise<BlsModule> {
  blsReady ??= blsInit()
  return blsReady
}

/** Verifies a BLS challenge signature produced by `signChallenge` in pegin-wasm. */
export async function verifyChallengeSignature(
  didPkHex: string,
  challenge: string,
  signatureHex: string,
): Promise<void> {
  const bls = await ensureBls()
  const pkBytes = hexToBytes(didPkHex)
  if (pkBytes.length !== 48) {
    throw new Error('DID public key must be 48 bytes')
  }
  const sigBytes = hexToBytes(signatureHex)
  if (sigBytes.length !== 96) {
    throw new Error('challenge signature must be 96 bytes')
  }
  const pk = bls.G1Element.from_bytes(pkBytes)
  const sig = bls.G2Element.from_bytes(sigBytes)
  const ok = bls.AugSchemeMPL.verify(pk, new TextEncoder().encode(challenge), sig)
  if (!ok) {
    throw new Error('challenge signature does not verify')
  }
}

function hexToBytes(hex: string): Uint8Array {
  const normalized = hex.startsWith('0x') ? hex.slice(2) : hex
  if (normalized.length % 2 !== 0) {
    throw new Error('invalid hex')
  }
  const out = new Uint8Array(normalized.length / 2)
  for (let i = 0; i < out.length; i++) {
    out[i] = Number.parseInt(normalized.slice(i * 2, i * 2 + 2), 16)
  }
  return out
}

File: crates/pegin-verify/src/modules/challenge/challenge.service.rs
//! Challenge signature verification (BLS `AugSchemeMPL` on DID key).

use chia_bls::{verify, PublicKey, Signature};

pub fn verify_challenge_signature(
    did_pk_bytes: &[u8],
    challenge: &str,
    signature_hex: &str,
) -> Result<(), String> {
    let pk_bytes: [u8; 48] = did_pk_bytes
        .try_into()
        .map_err(|_| "DID public key must be 48 bytes".to_owned())?;
    let pk =
        PublicKey::from_bytes(&pk_bytes).map_err(|e| format!("invalid DID public key: {e}"))?;
    let sig_bytes =
        hex::decode(signature_hex).map_err(|e| format!("invalid signature hex: {e}"))?;
    let sig_array: [u8; 96] = sig_bytes
        .try_into()
        .map_err(|_| "challenge signature must be 96 bytes".to_owned())?;
    let sig =
        Signature::from_bytes(&sig_array).map_err(|e| format!("invalid BLS signature: {e}"))?;
    if verify(&sig, &pk, challenge.as_bytes()) {
        Ok(())
    } else {
        Err("challenge signature does not verify".to_owned())
    }
}

#[cfg(test)]
mod tests {
    use chia_bls::{sign, SecretKey};

    use super::*;

    #[test]
    fn round_trip_challenge_verify() {
        let sk = SecretKey::from_bytes(&[9u8; 32]).expect("key");
        let pk = sk.public_key();
        let challenge = "login-nonce-abc";
        let sig = sign(&sk, challenge.as_bytes());
        verify_challenge_signature(&pk.to_bytes(), challenge, &hex::encode(sig.to_bytes()))
            .expect("verify");
    }
}
